File: other/binary_search/binary_search.cpp
#include <iostream>

using namespace std;

int iterative_bs(int n);
int recursive_bs(int n, int low, int high);

int array[] = {2, 3, 5, 7, 8, 10, 14, 20};
const int SIZE = 8;

int main() {

  cout << iterative_bs(14) << endl;
  cout << iterative_bs(100) << endl;
  cout << recursive_bs(14, 0, SIZE) << endl;
  return 0;
}

int recursive_bs(int n, int low, int high) {
  if (high < low) return -1;

  int mid = (low+high)/2;
  if(array[mid] < n) 
    recursive_bs(n, mid+1, high);
  else if(array[mid] > n)
    recursive_bs(n, low, mid-1);
  else
    return array[mid];
}

int iterative_bs(int n) {
  int low = 0, high = SIZE;

  while(high > low) {
    int mid = (high+low)/2;
    if (array[mid] < n) {
      low = mid+1;
    } else if (array[mid] > n) {
      high = mid-1;
    } else {
      return array[mid];
    }
  }
  return -1;
}

File: algorithms/sorting/binary_search.py
#!/usr/bin/python3

a = [13, 19, 21, 34, 48, 51, 56, 65, 67, 84, 88, 93]

def binary_search(a, low, high, value):
    if high < low:
        return None

    mid = (high+low)//2
    if a[mid] < value:
        return binary_search(a, mid+1, high, value)
    elif a[mid] > value:
        return binary_search(a, low, mid-1, value)
    else:
        return a[mid]

print(binary_search(a, 0, len(a)-1, 67))

def binary_search_iterative(a, value):
    low = 0
    high = len(a)-1
    while low <= high:
        mid = (low+high)//2
        if a[mid] < value:
            low = mid+1
        elif a[mid] > value:
            high = mid-1
        else:
            return a[mid]
    return None

print(binary_search_iterative(a, 34))